### PR TITLE
Added Type-Alias Support

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR adds support for type-aliases at the parser level (so it works for all languages).
It doesn't add logic for adding `using` directives in the generated code.

### CODE REVIEW

File: cpp/src/Slice/Grammar.y
Comment (pepone) [inline cpp/src/Slice/Grammar.y:1572]: Same here, we can get rid of the nested if

`if (!interface && auto alias = ...`

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (pepone) [inline cpp/test/Slice/errorDetection/TypeAlias.ice:57]: I don't see how this will work, the only generated class is C0, what is the purpose of applying metadata to the alias?

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (pepone) [inline cpp/test/Slice/errorDetection/TypeAlias.ice:62]: I would expect this to be an error

**Review by pepone — COMMENTED** [review]
I find the `unwrap/wrap/mergeMetadataInPlace` pattern too verbose wondering if we can come up with a simpler implementation.
```
    if (auto alias = TypeAliasPtr::dynamicCast(unwrapIfOptional(type)))
    {
        mergeMetadataInPlace(metadata->v, alias->typeMetadata());
        type = rewrapIfOptional(alias->underlying(), type);
    }
```

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1069]: Absolutely! I separated them to keep the variable initialization readable, but I don't really care either way.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1071]: This check is specifically for parsing:
`class Bar extends FooAlias`
We forbid any metadata to be specified on `FooAlias`, since that would over-complicate inheritance.
So here, we check if any metadata was specified on the alias, and if there was, we raise an error.

So to answer your question: None can appear here! And this check is the reason why it can't.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1572]: Yep, fixed same as above.

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/errorDetection/TypeAlias.ice:57]: From a Slice perspective, you're right, these aliases are useless.
From a test perspective though, these are here to make sure that aliases with metadata are parsed correctly and don't raise any unexpected errors. It's just for testing purposes.

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/errorDetection/TypeAlias.ice:62]: The way I've written type-aliases is that any metadata can be applied in any type-alias, and errors are only reported when the type-alias is actually used. So if you used `I0AliasMetadata` somewhere, it would raise an error for invalid metadata.

This seemed more flexible, since metadata validity can be context dependent.
Plus, right now, all our metadata validation is performed _after_ the parsing stage (usually by metadata visitors), when type-aliases are already resolved. It seems simpler just to use our existent metadata validation, instead of adding another validation step in the parsing stage.

**Comment by InsertCreativityHere** [comment]
I agree that the pattern is over complicated.
Ideally, I'd like to make optional into a boolean field on type, instead of a separate wrapper class... Then we wouldn't need to have this rewrap logic at all. But that's a somewhat large change I didn't want to make in the middle of this one.
I'll blog about it on Monday to see if anyone has a compelling argument for keeping the wrapper class and go from there.

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (pepone) [reply cpp/test/Slice/errorDetection/TypeAlias.ice:62]: maybe we can delegate the validation to the type being aliases and the validation should decide if the metadata is valid for the type and for the type alias, I think [amd],[prtected] only make sense to accept it as metadata for the type but not for the alias. Anyway this is something we can rework later.

Comment (InsertCreativityHere) [reply cpp/test/Slice/errorDetection/TypeAlias.ice:62]: Yes, I agree that `[amd]` and `[protected]` are totally bogus here! I don't have any issue with moving the validation onto the alias, but it'll probably have to wait until after I get back to my metadata-related changes... Right now our metadata validation just isn't very powerful or flexible.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1069]: Nevermind, I spoke too fast and forgot I was initializing a variable in the if statement.
I don't believe C++ allows you to both evaluate an expression and declare a variable in the same if statement... Unless there's some super fancy C++20 syntax I'm unaware of.

Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y:1069]: The feature is called "if statement with initializer" and is available as of C++17:
https://en.cppreference.com/w/cpp/language/if

```
if (auto alias = xxx; alias && !cl)
{
   ... 
}
```

Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1069]: Yes, what I meant to say is there's no way to do it AND keep the same order (checking `cl` first). I could change the order, but it seemed wasteful to always perform this cast when 99% of the time, `!cl` is false, and so the cast is pointless to do.
If you really want, I can change it, but the current 2-if version is more efficient and in my opinion also more readable than
`if (auto alias = TypeAliasPtr::dynamicCast(types.front()); alias && !cl)`
So even if it would save us 3 lines of code... I'd personally prefer to leave it as is.

Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1069]: For reference, this is the current code in my PR:
```
        if (!cl)
        {
            if (auto alias = TypeAliasPtr::dynamicCast(types.front()))
            {
                ...
            }
        }
```
vs the other way of writing it:
```
        if (auto alias = TypeAliasPtr::dynamicCast(types.front()); alias && !cl)
        {
            ...
        }
```

Comment (pepone) [reply cpp/src/Slice/Grammar.y:1069]: The syntax is a bit verbose, fine with me if you prefer to keep it as a nested if

File: cpp/src/Slice/SliceUtil.cpp
Comment (bernardnormier) [inline cpp/src/Slice/SliceUtil.cpp:949]: Each metadata is a string, not a map entry, so what does precedence mean here?

For example, if one metadata is cs:type(Foo) and the other metadata is cs:type(Bar), both would be added. I think in this case you want to add only one, but this requires more structure to the metadata.

File: cpp/src/Slice/Parser.cpp
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:1194]: This code does not make sense. We should not allow type aliases for exceptions at all.

File: cpp/src/Slice/Grammar.y
Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y:1071]: This check is in the wrong spot. 
When you define a typealias for a class, the Parser should disallow any metadata. So when you call typeMetadata() on a typealias for a class, you always get an empty list.

File: cpp/src/Slice/Grammar.y
Comment (bernardnormier) [inline cpp/src/Slice/Grammar.y:1534]: Here, the check is in the wrong spot. The empty metadata list should be enforced when the alias is created.

**Review by bernardnormier — COMMENTED** [review]
Like Jose, I don't like the unwrap /  merge / rewrap pattern. I also think the merge is incorrect (see comment).

For example, if I define:

```
using DistanceInMeter = int;
using DistanceList = sequence<DistanceInMeter>;
```

It would be nice if the corresponding C++ code was:
```
using DistanceInMeter = int32_t;
using DistanceList = std::vector<DistanceInMeter>;
```

But you're erasing the type alias from the sequence element type (etc). so generating the above is not possible.
I think it would be simpler and more correct to keep the type aliases everywhere, and allow the Slice compilers to use the type alias info (in particular the identifier).

File: cpp/test/Slice/escape/Key.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/escape/Key.ice:67]: Fair enough! I'll add it back in.

File: cpp/src/Slice/SliceUtil.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:949]: Even if it's a list and there can be repeats, often we will still only use the first entry we find. For instance, if you write
`[cs:type(Foo)] [cs:type(Bar)] ...`
Obviously a member can't have both types at the same time. In this case `Foo` _has precedence_ here, meaning that is the actual type that will be used. It has precedence because it will be first in the list.
This is what my comment is alluding to.

But you're also right, that we would need extra structure to do a proper merge like this... I forgot that my metadata changes never got merged in to master. I'll change these for now until I do get to merge them in.

File: cpp/src/Slice/Parser.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:1194]: True, I got carried away while adding the code for classes and structs. I'll remove this.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1534]: This check is specifically for inheritance, like while parsing:
`interface Foo : BarAlias`
In this specific case, obviously we don't allow metadata here:
`interface Foo : [metadata] Bar    // bogus`
And so the check makes sure the list is empty to enforce this.

But we don't know what context the typealias will be used in, and so the check has to be delayed until we have that context (after type-alias creation).
Or are you saying that we should disallow type-aliases of interfaces and classes to ever have metadata in any circumstance? (lumping in the above comment with this one)

**Comment by InsertCreativityHere** [comment]
I agree that only type and type-like metadata should be allowed, but right now sorting metadata into type and non-type metadata isn't something we have the structure to do. This is what my metadata validation changes were going to be for before I put them on the back-burner. Once I get back around to finishing those changes, this code will be updated to take advantage of them and only allow type metadata.

As for the erasure of type-alias information, this is a natural feature of the type-aliases being transparent, like I said on my blog posts. I can make them non-transparent, but this will require more significant changes to the parser... the issue is that we use casting to determine the types of slice entities while parsing, and if we leave type-aliases as first class types (instead of transparently unwrapping them), our current setup that relies almost entirely casting completely falls apart, since
```
class Foo;
using Bar = Foo;
```
the type of `Bar` here isn't a `ClassDef`, it'll be a `TypeAlias`.

I already know how to implement all the changes to allow for non-transparent type-aliases, since I was going to have to make them anyways for removing forward declarations, I was just putting them off until then since I thought everyone was okay with the type-aliases being transparent like this. I'll close this PR for now, and explain my proposal on a blog post.

File: cpp/src/Slice/Grammar.y
Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y:1534]: > Or are you saying that we should disallow type-aliases of interfaces and classes to ever have metadata in any circumstance? (lumping in the above comment with this one)

Yes. And more specifically:
a) the only metadata allowed in a typealias definition is type-metadata such as cs:type(), cpp:type() etc. Non-type metadata such as [amd] or [protected] or [cpp:const] (...) are not allowed.
b) for class and interface typealiases, we don't allow any metadata at all

Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1534]: As of current metadata, B is already included in A, since there isn't any type metadata that can be applied to interfaces or classes that I'm aware of!

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y]: I moved the checks so now defining metadata on a type-alias of a class or interface will report the error at type-alias creation, instead of where it's used.

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/errorDetection/TypeAlias.ice]: I agree that type-like metadata should be an error, but right now we don't have the ability to distinguish between different types of metadata. I'll tighten these checks once I've had a chance to add some structure to how we handle metadata.

File: cpp/test/Slice/errorDetection/TypeAlias.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/errorDetection/TypeAlias.ice]: Same as above, I'll rework this once we have stronger metadata logic so we can filter type and non-type metadata.

File: cpp/test/Slice/escape/Key.ice
Comment (InsertCreativityHere) [reply cpp/test/Slice/escape/Key.ice:67]: Added it back in!

File: cpp/src/Slice/SliceUtil.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp]: I fixed the implementation of these so even repeated metadata is copied over, but I left the helper functions, since even though they're short, we use their logic in many places, and they're much less verbose than working with the lists directly.

File: cpp/src/Slice/Parser.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp]: This was removed.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y]: I moved the check to when the type-alias is created, and added a check that forbids metadata to be specified on aliases of classes and interfaces.
It's on line 1685 of Grammar.y (I don't think I can link to lines in PRs on Github)

**Comment by InsertCreativityHere** [comment]
For the time at hand, we've decided transparent aliases are good enough. I'll revisit them later once I've laid out some other changes to help with implementing opaque aliases in a clean manner.

I haven't added in any code for forbidding non-type metadata on aliases, since right now we don't distinguish between different types of metadata at the parsing level. Soon, I'll add better logic to our metadata handling to enable this kind of fine-grained filtering.

I can't think of any other way to replace the unwrap/merge/rewrap pattern. Both TypeAlias and Optional are wrapper classes, and Optional **HAS** to be the outermost wrapper. I can't think of any way to do this without a rewrap. In the future we can reconsider changing how we've implemented Optional, but that's outside the scope of this PR.

I did move the logic to a `resolveAlias` helper function, to make it less verbose at least.

File: cpp/src/Slice/Grammar.y
Comment (pepone) [inline cpp/src/Slice/Grammar.y:1700]: Isn't this an error?

File: cpp/src/Slice/Parser.cpp
Comment (pepone) [inline cpp/src/Slice/Parser.cpp:2069]: Should we assert that type is not an alias itselft?

File: csharp/test/Slice/alias/Client.cs
Comment (pepone) [inline csharp/test/Slice/alias/Client.cs:11]: I think it will be better if we have a test that actually exercises these aliases, checking that the operation has the expected return type doesn't ensure they work, something might be wrong with the marshaling or unmarshaling when using type alias and we will not notice it.

**Review by pepone — APPROVED** [review]
Looks good, but added some comments

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1700]: Yes, but we report errors like this when the type resolution fails, so by the time we get here, we've already issued one.

However, we still need to explicitly set `$$ = nullptr` here, because if Bison reaches the end of a statement and `$$` hasn't been set, it will implicitly set it to `$$ = $1`, which would be the `ICE_USING` token here, and which we don't want to pass up the parsing tree.

File: cpp/src/Slice/Parser.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:2069]: Yes, it couldn't hurt. I've added an assert here.

File: csharp/test/Slice/alias/Client.cs
Comment (InsertCreativityHere) [reply csharp/test/Slice/alias/Client.cs:11]: Fair enough, I'll add more to the test for actually using them.

File: cpp/src/Slice/SliceUtil.cpp
Comment (bernardnormier) [inline cpp/src/Slice/SliceUtil.cpp:955]: These two functions are not "merging" the list<string> but concatenating them. See list::merge.
I think a single concatMetadata or appendMetadata function would be sufficient.

File: cpp/src/Slice/Grammar.y
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:1437]: This is required for when there's metadata specified on the return type, and on the operation as a whole:
```
using Foo = [cs:type(Hello)] Bar;
...
[cs:type(Something)] Bar myOp();
```
So here, the operation metadata (`cs:type(Something)`) has to be merged into the metadata already on `Bar`.

I'll add a comment about it though!

File: cpp/src/Slice/SliceUtil.cpp
Comment (bernardnormier) [inline cpp/src/Slice/SliceUtil.cpp:679]: I find the signature of this resolve function with 2 in-out parameters inelegant.

```
(TypePtr, StringList) resolveAlias(const TypePtr type&, const StringList& metadata);
```
looks better.

File: cpp/src/Slice/Util.h
Comment (bernardnormier) [inline cpp/src/Slice/Util.h:73]: Type aliases are "transparent", meaning when we generate code for a sequence<T> or T member, where T is an alias, we only see the actual underlying type. As a result, `unwrapIfAlias` is not really useful. When generating code for aliases, surely we don't need to auto unwrap them.

File: cpp/src/Slice/Util.h
Comment (bernardnormier) [inline cpp/src/Slice/Util.h:74]: Same for unwrapType.

Comment (InsertCreativityHere) [reply cpp/src/Slice/Util.h:74]: I removed `unwrapType`, it was only used in one place, and only called the other two unwrap functions under the hood anyways.

File: cpp/src/Slice/Util.h
Comment (InsertCreativityHere) [reply cpp/src/Slice/Util.h:73]: Unfortunately there are places where this is useful. Not many, there are some places in the language-specific compilers where we do type-lookup manually (by calling `unit->lookupXXX()`), which will get the actual type-alias and won't perform any automatic unwrapping (we keep the original type-aliases around in case we want to generate `using` statements for them).

I removed a few of these places in my PR, but there were others that were less trivial to fix, where these helpers were useful to perform the check and unwrap in a single line.

File: cpp/src/Slice/SliceUtil.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:955]: Fair enough, does `combineMetadata` sit better? I'd like to avoid `appendMetadata`, since even though that's what it does now, it's only because we lack advanced metadata logic. When I add it (which I plan to do somewhat soon), it will do more than just simply appending one list to another. So `appendMetadata` will only be a temporarily correct name.

File: cpp/src/Slice/SliceUtil.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:679]: Normally, I'd agree that in-out parameters are 'more risky', but everywhere we use this method, we want to directly overwrite something in place. Making the parameters const references and instead returning the outputs would just be adding a pointless step to this, since we're going to overwrite the original values anyways.

And while you think it makes the function signature look nicer, it would make the implementation and everywhere where we use it worse. For instance, this:
```
resolveAlias(taggedDef->type, taggedDef->metadata);
```
would become this:
```
tie(taggedDef->type, taggedDef->metadata) = resolveAlias(taggedDef->type, taggedDef->metadata);
```
We can't use structured binding, because you can only bind to new variables, not existing ones, like we have to here.

The other advantage of not returning anything is that when the provided type isn't an alias (which is often), the method can just immediately return and nothing is changed.
If we used the tuple return type like you suggest, then the method would instead return an exact copy of the inputs, and in the grammar we would overwrite the type and metadata with an exact copy of themselves.

Comment (bernardnormier) [reply cpp/src/Slice/SliceUtil.cpp:673]: Then I suggest to change the name of the function to for example `unalias`.

File: cpp/src/Slice/SliceUtil.cpp
Comment (bernardnormier) [reply cpp/src/Slice/SliceUtil.cpp:949]: You can fix the function name once you change its implementation to have different semantics.

Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp:949]: Alright, I renamed it to 'appendMetadata' for the time at hand.

File: cpp/src/Slice/SliceUtil.cpp
Comment (InsertCreativityHere) [reply cpp/src/Slice/SliceUtil.cpp]: `unalias` works for me. I changed it.

**Pull request merged**
